### PR TITLE
Sdks 1931 fix memory leak

### PR DESCRIPTION
**Pull request by vahancouver**
# JIRA Ticket

[SDKS-1931](https://bugster.forgerock.org/jira/browse/SDKS-1931)

# Description

Change to capture reachabilityMonitor as weak in reachabilityMonitor.monitoringCallback to avoid memory leaks. 
Add Unit tests to prove this

# Definition of Done Checklist:

- [x] Acceptance criteria is met.
- [x] All tasks listed in the user story have been completed.
- [x] Coded to standards.
- [x] Ensure backward compatibility.
- [ ] API reference docs is updated.
- [x] Unit tests are written.
- [ ] Integration tests are written.
- [ ] e2e tests are written.
- [ ] Functional spec is written/updated.
- [ ] Example code snippets have been added.
- [x] Change log updated.
- [ ] Documentation story is created and tracked.
- [ ] Tech debts and remaining tasks are tracked in separated ticket(s).

### CODE REVIEW

File: FRAuth/FRAuth/Device/NetworkCollector.swift [hunk 1]
```diff
@@ -28,9 +28,9 @@ public class NetworkCollector: DeviceCollector {
         if let reachabilityMonitor = NetworkReachabilityMonitor() {
             reachabilityMonitor.startMonitoring()
             
-            reachabilityMonitor.monitoringCallback = { (status) in
-                result["connected"] = reachabilityMonitor.isReachable
-                reachabilityMonitor.stopMonitoring()
+            reachabilityMonitor.monitoringCallback = { [weak reachabilityMonitor] (status) in
+                result["connected"] = reachabilityMonitor?.isReachable
+                reachabilityMonitor?.stopMonitoring()
```
[the remark this answers is not in the document]
Comment (vahancouver) [reply FRAuth/FRAuth/Device/NetworkCollector.swift:33]: But we still want completion(result) line to work. that would invalidate the whole closure. Plus it's redundant as we know the compiler will remove the object if there is no mutual strong reference.

[the remark this answers is not in the document]
Comment (vahancouver) [reply FRAuth/FRAuth/Device/NetworkCollector.swift:33]: 2. I tried to write a test for this class to exactly test this, but it's impossible as we don't have reference to `reachabilityMonitor` . It's created inside the method. I think my test covered what we wanna test. 

Honestly this is the most classic retain cycle with closures, and capturing the instance as weak is the way to solve it.  It's exactly the same what we do with self when self owns the closure.

[the remark this answers is not in the document]
Comment (vahancouver) [reply FRAuth/FRAuth/Device/NetworkCollector.swift:33]: @jeyanthanperiyasamy you had a fix for the memory leak. I wonder what was your fix?

[the remark this answers is not in the document]
Comment (jeyanthanperiyasamy) [reply FRAuth/FRAuth/Device/NetworkCollector.swift:33]: we are good here, we talked about few approaches and testing practices. i think the fix looks good to me.

**Review by jeyanthanperiyasamy — APPROVED** [review]

**Review by spetrov — APPROVED** [review]
I reviewed and tested the fix. 
Looks good to me! 👍🏻

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,4 +1,7 @@
 # Version 3.3.1
+## [Unrelease]
```
Comment (george-bafaloukas-forgerock) [inline CHANGELOG.md:2]: Unreleased - typo ;)

File: FRAuth/FRAuth/Device/NetworkCollector.swift (hunk 1 above)
[the remark this answers is not in the document]
Comment (george-bafaloukas-forgerock) [reply FRAuth/FRAuth/Device/NetworkCollector.swift:33]: The fix looks good. I would be interested to see if there any other closures around this functionality (other collectors) that need this "weakening" as well

**Review by george-bafaloukas-forgerock — APPROVED** [review]
Looks good. There is a typo in the change log

File: FRAuth/FRAuth/Device/NetworkCollector.swift (hunk 1 above)
[the remark this answers is not in the document]
Comment (jeyanthanperiyasamy) [reply FRAuth/FRAuth/Device/NetworkCollector.swift:33]: good point @george-bafaloukas-forgerock .    vahan  @vahancouver  can you swing other collectors as well and see is there anything we miss ?

[the remark this answers is not in the document]
Comment (vahancouver) [reply FRAuth/FRAuth/Device/NetworkCollector.swift:33]: @george-bafaloukas-forgerock @jeyanthanperiyasamy I double checked and other Collectors don't have this issue. In fact this is not really a collector issue, it's NetworkReachabilityMonitor issue (NetworkCollector happened to use it)

**Pull request merged**